### PR TITLE
KAFKA-12375: fix concurrency issue in application shutdown

### PR DESCRIPTION
**Pull request by wcarlson5**
enforce rebalance is now used in a thread safe way

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 1]
```diff
@@ -663,7 +666,6 @@ public void sendShutdownRequest(final AssignorError assignorError) {
         log.warn("Detected that shutdown was requested. " +
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: nit: move this to where we call `enforceRebalance` as well

**Review by ableegoldman — APPROVED** [review]
Thanks for the quick fix! This should be cherrypicked to 2.8 cc @vvcephei

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: done

**Review by cadonna — APPROVED** [review]
Thank you @wcarlson5 !
LGTM!
I had the same idea! 🙂

**Comment by ableegoldman** [comment]
One unrelated test failure: `kafka.server.ListOffsetsRequestTest.testResponseIncludesLeaderEpoch`

Going to merge this now

**Comment by ableegoldman** [comment]
Merged to trunk and cherrypicked to 2.8 cc @vvcephei

**Pull request merged**